### PR TITLE
Improved ranking + confidence rating for suggestions

### PR DESCRIPTION
**Pull request by fabrichter**
Added a ML model for ranking suggestions for spelling corrections, trained on user-submitted data.
Extended JSON format of response accordingly with two new fields:

```json
"replacements": [{"value":"mistake","autoCorrect":true,"confidence":0.9723154}]
```
`autoCorrect` is set to true only when confidence is above a given threshold that gives sufficient precision.

For now, this is only enabled in English and as an A/B test (enable via `abTest=SuggestionsRanker` in the server configuration file).

### CODE REVIEW

File: languagetool-core/src/main/java/org/languagetool/rules/spelling/SymSpellRule.java [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+import org.languagetool.rules.spelling.symspell.implementation.SymSpell;
+
+import java.io.*;
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+public class SymSpellRule extends SpellingCheckRule {
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/rules/spelling/SymSpellRule.java]: Please add a javadoc describing the differences to the other spell rules

File: languagetool-core/src/main/java/org/languagetool/rules/spelling/SymSpellRule.java [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+      });
+      speller.commitStaged(stage);
+      return speller;
+    } else {
+      return null;
+    }
+  }
+
+  protected static SymSpell initDefaultDictSpeller(Language lang) {
+    SymSpell speller = new SymSpell(100000, 3, -1, 0);
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/rules/spelling/SymSpellRule.java]: Maybe using a named constant here for 100000?

File: languagetool-core/src/main/java/org/languagetool/rules/spelling/SymSpellRule.java [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+  public String getDescription() {
+    return "Spell checking rule using SymSpell algorithm";
+  }
+
+  @Override
+  public RuleMatch[] match(AnalyzedSentence sentence) throws IOException {
+    List<RuleMatch> matches = new ArrayList<>();
+    Set<String> ignoredWords = ignoredWordsCache.getUnchecked(language);
+    for (AnalyzedTokenReadings token : sentence.getTokensWithoutWhitespace()) {
+      if (token.isSentenceStart() || token.isSentenceEnd() || token.isNonWord())
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/rules/spelling/SymSpellRule.java]: Really token.isSentenceEnd()? Aren't there cases where the sentence ends with a word?

File: languagetool-core/src/main/java/org/languagetool/rules/spelling/suggestions/SuggestionsChanges.java [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public class SuggestionsChanges {
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/rules/spelling/suggestions/SuggestionsChanges.java]: Please add javadoc

File: languagetool-core/src/main/java/org/languagetool/rules/spelling/symspell/implementation/ChunkArray.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
+package org.languagetool.rules.spelling.symspell.implementation;//        MIT License
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/rules/spelling/symspell/implementation/ChunkArray.java]: line break

File: languagetool-core/src/main/java/org/languagetool/rules/spelling/symspell/implementation/EditDistance.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
+package org.languagetool.rules.spelling.symspell.implementation;//        MIT License
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/rules/spelling/symspell/implementation/EditDistance.java]: line break

File: languagetool-core/src/main/java/org/languagetool/rules/spelling/symspell/implementation/SuggestItem.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
+package org.languagetool.rules.spelling.symspell.implementation;//        MIT License
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/rules/spelling/symspell/implementation/SuggestItem.java]: line break

File: languagetool-core/src/main/java/org/languagetool/rules/spelling/symspell/implementation/SymSpellDemo.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
+package org.languagetool.rules.spelling.symspell.implementation;//        MIT License
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/rules/spelling/symspell/implementation/SymSpellDemo.java]: line break

File: languagetool-core/src/main/java/org/languagetool/tools/RuleMatchesAsJsonSerializer.java [hunk 9]
```diff
@@ -45,7 +47,8 @@
   private static final String START_MARKER = "__languagetool_start_marker";
 
   private final JsonFactory factory = new JsonFactory();
-  
+  private static final Logger logger = LoggerFactory.getLogger(RuleMatchesAsJsonSerializer.class);
```
Comment (danielnaber) [inline languagetool-core/src/main/java/org/languagetool/tools/RuleMatchesAsJsonSerializer.java]: not used, is it?

File: languagetool-core/src/main/java/org/languagetool/rules/spelling/SpellingCheckRule.java [hunk 10]
```diff
@@ -93,16 +97,94 @@ public SpellingCheckRule(ResourceBundle messages, Language language, UserConfig
...
+          // could not rank for some reason
+        } else {
+          if (userCandidates.size() == 0) {
+            match.setAutoCorrect(ranker.shouldAutoCorrect(defaultSuggestions));
+            match.setSuggestedReplacementObjects(defaultSuggestions);
+          } else {
+            List<SuggestedReplacement> combinedSuggestions = new ArrayList<>();
+            for (String wordFromUserDict : userCandidates) {
+              SuggestedReplacement s = new SuggestedReplacement(wordFromUserDict);
+              s.setConfidence(0.0f);
```
[the remark this answers is not in the document]
Comment (fabrichter) [reply languagetool-core/src/main/java/org/languagetool/rules/spelling/SpellingCheckRule.java]: done

**Pull request merged**